### ISSUE TITLE
Proposed API to move us away from Promises+Async for output properties.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [pulumi/pulumi#808]
TLDR: I have a proposal on how we can do synchronous resource creation that has a reasonably non-sucky user programming model, while also enabling efficiency and parallelism.  Please review and give me your thoughts.  Sorry for the long expository paragraphs, but i couldn't think of a better way to present it and explain it all.  Happy to discuss in person or any other medium.  Thanks!

--

I've whipped up a brief proposal that i think will allow us to move away from the async/promise based approach we have to input/output properties (i.e. computed) to a more synchronous approach that will hopefully be simpler and more straightforward for customers to use, while still keeping things flexible on our end to achieve good performance.

I intend to then implement this API; initially opting for a less efficient mechanism, while keeping the door open for us to then exchange that with a more efficient approach later on.  

--

The primary areas i want to focus on are:
1. Making it easy for us to track dependencies in resources.
2. Making is simple (and non-async) for users to actually work with the *value* of properties.
3. Ensuring that users can work and transform the *values* of properties, without sacrificing our ability to track dependencies or operate efficiently (i.e. in parallel).

To that end, i have come up with the following APIs (note: all names non-final, suggestions welcome).

```ts
interface Dependency<T> {
     get(): T;
     apply<U>(func: (t: T) => U): Dependency<U>;
     ids(): number[];
}
```

A ```Dependency<T>``` represents the ability to retrieve and transform data in a *deferred* manner.  In general, all output properties of a resource would be ```Dependency<SomeType>```, while all inputs to a resource would be ```SomeType | Dependency<SomeType>```.  This allows outputs of a resource to be easily passed, unmodified, into the inputs of another resource, allowing dependency tracking information to be captured easily in the common case.

First, let's look at 'ids'.  For the purpose of this review the actual type of 'ids' is not important.  It could be a number, or string, or any object as long as it provides enough information for Pulumi to actually track what resource this dependency property value came from.  Importantly, it likely should not be public either, but rather will just be data under the covers that is shuttled around to allow resource creation to then encode what dependencies it has.  However, for the purposes of this review, i've made it publicly visible so it's easy to remember that it is there.  In practice, "registerResource" will use this to pass information to the Resource Manager to help create an explicit and accurate dependency graph between resources.

Now, it's not always possible to just pass an output property from one resource in as input of another.  To that end, two methods on ```Dependency<T>``` are provided to make transformation of property values possible.  The first is ```get(): T```.  As the signature implies, it synchronously acquires and gives you back the resource.  It is extremely simple to use in code, but has the downside of "losing" the information the Dependency object actually tracks.  If customers use this method (which i expect they will), transforms the result, and then passes that transformed value to a new resource, we will lose the nice explicit dependency tracking information.  Because of that, our current thinking is that "get()" will go back and tell the Resource Manager it was called, at which point the resource manager will then consider all later resource creation to be implicitly dependent on the Resource that exposed this property.  This is very heavyweight and efficiency sapping, but it's something we believe we can visualize to users in the future, and help them move away from if it is an actual problem***.  In general, we believe that this will not actually be problematic the majority of the time for users.

Because of these issues with "get()", we also expose another method ```apply<U>(func: (t: T) => U): Dependency<U>``` that helps address the problems for when our users (or ourselves) need it.  As the name implies, ```apply``` allows a user to 'apply' a transformation to the deferred value of a Dependency property (for example, manipulating numeric values, or potentially transforming strings).  Importantly, the result of ```apply``` is itself a ```Dependency```, allowing it to them be passed as an input property to another Resource.  In the implementation ```Dependency``` will expose the same ```ids()``` as the original ```Dependency``` it was created out of.  With this, the resource being created can then accurately record the original Resource(s) that it depends on for the Resource Manager.  Because of this, any number of transformations can be 'applied' without losing dependency information.

Now, in practice, ```apply``` will be sufficient for a great many transformation cases.  However, it's not always sufficient, as there may be cases that need the results of several output properties in order to create a single input property to another resource.  While it is possible in those circumstances to use ```.get()``` on all the Dependencies, all the problems with ```get()``` will then be incurred.  To alleviate this problem, a set of combinators are also provided like the following:

```ts
// D<T> is just shorthand for Dependency<T>
export function combine<T1, T2>(d1: D<T1>, d2: D<T2>): D<[T1, T2]>;
export function combine<T1, T2, T3>(d1: D<T1>, d2: D<T2>, d3: D<T3>): D<[T1, T2, T3]>;
// etc.
```

What the first sig says is "Given two Dependencies of two deferred values, combine them into a single Dependency of a single deferred tuple containing both those values".  The benefit of that, is that now ```apply``` can be called on that new ```Dependency```, which will have access to both values which it can then transform as it wants.  *Importantly* this new Dependency encodes all the ```ids``` information from *both* starting Dependency objects.  So if this new Dependency is passed to a newly created resource, that resource can accurately convey to the Resource Manager all the *original* resources it depends on.  

Here's an example of ```combine``` in practice:

```ts
// 'command' will be a Dependency<string[]>.  
// stacks.table.name, updates.table.name, kmsKey.arn are all Dependency<string>
const command = pulumi.combine(stacks.table.name, updates.table.name, kmsKey.arn).apply(
    // This function is strongly typed.  The types of the 3 arguments are all properly inferred
    // from the original dependency objects passed in.
    ([stackTableName, updateTableName, arn]) => {
        const comment: string[] = ["ppcd", "-v=10", "--logtostderr", "-token", accessToken,
                "-aws.stacks", stackTableName,
                "-aws.updates", updateTableName,
                "-aws.keyid", arn];

            if (debug) {
                command.push("--enable-profiling");
            }

            return command;
    });

    const ppcdContainer = {
        build: "../build-context",
        ports: [{ port: 8080 }],
        command: command,  // command is used here, no dependency info is lost.
        memoryReservation: 1024,
        volumes: [{ containerPath: dockerSock, sourceVolume: new cloud.HostPathVolume(dockerSock) }],
    };
```

This works because the ```Container type``` needs to take a ```string[] | Dependency<string[]>```.  In this example, we provide ```Dependency<string[]>``` to enable as much Dependency information to actually be encoded so that our Resource Manager has extremely precise information, and thus can record this accurately.  

Note that this approach is certainly less ergonomic and less pleasant than using "get()".  But for advanced scenarios, and people who need perf, it does provide an appropriate mechanism to do things properly.

--

*** We will also likely want to encourage users to avoid 'get()' when possible and to just pass dependencies along, or do simple transformations with 'apply'.  This encouragement should come through user education, as well as potential visualization mechanisms.  For example, we can envision in the future that the user dashboard shows them what happened when their resources were being created, as well as showing resource dependencies.  ```get()``` calls would appear as very noticeable choke points where parallelization and decoupling stopped, and the whole rest of the world (after the ```get()``` point) became tightly dependent.

--

This API approach then enables the following set of work item steps that we can take over the next milestones.

1. Make *whole* resource creation *fully* synchronous. i.e. when the NodeJS side tries to create a resource, it all happens synchronously, and all output properties will be fully realized when the call completed.  Note: this will only be a temporary state that will be changed later on this list.  This can be done by having our JS code call into a C++ entrypoint (similar to how we do closure computation), which then uses synchronous C++ grpc to call the actual Resource Manager.  This allows us to actually unsyncify all our resource generation code, and all our resource property code.  Note that at this point ```get()``` won't do anything interesting as all output properties will be known once the resource is created.
2. Switch to using ```Dependency<T>``` (including removing Computed/ComputedVal) across the whole stack.
3. Have resource creation collect the ```ids()``` for their input properties and send that over to the Resource Manager.  Note: what the Resource Manager does with this is out of scope for this issue.
4. Have ```get()``` send a synchronous grpc message (again, using C++ rpc), along with the ```ids()``` of that Dependency object, to let the Resource manager know that all further resources depend on it.  Note: what the Resource Manager does with this is out of scope for this issue.
5. Eventually, make resource creation *partially* synchronous.  Specifically, the call to make the resource to the Resource Manger will happen synchronously, however, the resource object will be returned immediately, while the Resource Manager (written in Go) itself is asynchronously actually making it (i.e. calling into AWS and whatnot).   ```get()``` calls will then call back to the Resource Manager to wait for that resource to finish being created (ideally it may have already happened by that point), and to acquire the actual output property values.

Note that '5' is what enables parallelism and efficiency, but *importantly* does not require any API changes once we're past '2'.  To get the best behavior and experience, it will be important that people avoid using ```get()``` as much as possible.  In an ideal world, where one only passes dependencies along, and only transforms using ```apply``` (or ```combine+apply```), this should provide maximal parallelism.  This does require though that any input properties always allow for ```Dependency<T>``` to allow users to not have to realize the value of the property and stop-the-world while waiting to compute it.
